core/src/lib/channel.h

```diff
@@ -0,0 +1,349 @@
+/*
+   BAREOS® - Backup Archiving REcovery Open Sourced
+
+   Copyright (C) 2023-2023 Bareos GmbH & Co. KG
+
+   This program is Free Software; you can redistribute it and/or
+   modify it under the terms of version three of the GNU Affero General Public
+   License as published by the Free Software Foundation and included
+   in the file LICENSE.
+
+   This program is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+
+#ifndef BAREOS_LIB_CHANNEL_H_
+#define BAREOS_LIB_CHANNEL_H_
+
+#include <condition_variable>
+#include <optional>
+#include <vector>
+#include <utility>
+#include <variant>
+
+#include "lib/thread_util.h"
+
+namespace channel {
+// a simple single consumer/ single producer queue
+// Its composed of three parts: the input, the output and the
+// actual queue data.
+// Instead of directly interacting with the queue itself you instead
+// interact with either the input or the output.
+// This ensures that there is only one producer (who writes to the input)
+// and one consumer (who reads from the output).
+
+struct failed_to_acquire_lock {};
+struct channel_closed {};
```

Comment:
I have the impression that in the C++ standard world the custom is to declare a type like `struct X_t { explicit X_t() = default; };` and then have `constexpr X_t X{}` as the only member.

Reply:
Yes that is what standard C++ often does.  I try to avoid this since all types ending in `_t` are reserved by posix.
I do not mind changing this to `channel_closed_t` tough.   What do you think ?